src/math/bvh.js

```diff
@@ -0,0 +1,451 @@
+        this.bvhNode[rightChildIdx].leftFirst = i;
+        this.bvhNode[rightChildIdx].triCount = node.triCount - leftCount;
+        node.leftFirst = leftChildIdx;
+        node.triCount = 0;
+        this.UpdateNodeBounds(leftChildIdx);
+        this.UpdateNodeBounds(rightChildIdx);
+
+        // Recurse into each of the child nodes
+        this.Subdivide(leftChildIdx);
+        this.Subdivide(rightChildIdx);
+    }
+
+    /**
+     * Builds the BVH
+     *
+     * @ignore
+     */
+    BuildBVH() {
+        const N = this.triangles.length;
+        this.triIdx = Array.from({ length: N }, (x, i) => i);
+        this.bvhNode = Array.apply(null, Array(N * 2)).map(function () {
+            return new BVHNode();
+        });
+        const rootNodeIdx = 0;
+        this.nodesUsed = 1;
+
+        for (let i = 0; i < N; i++) {
+            this.triangles[i].centroid = new Vec3();
+            this.triangles[i].centroid.add(this.triangles[i].vertex0);
+            this.triangles[i].centroid.add(this.triangles[i].vertex1);
+            this.triangles[i].centroid.add(this.triangles[i].vertex2);
+            this.triangles[i].centroid.mulScalar(1 / 3);
+        }
+
+        const root = this.bvhNode[0];
+        root.leftFirst = 0;
+        root.triCount = N;
+        this.UpdateNodeBounds(rootNodeIdx);
+        this.Subdivide(rootNodeIdx);
+    }
+
+    /**
+     * Refits the BVH
+     *
+     * @param {Array} [triangles] - The triangles to refit the BVH on
+     * @ignore
+     */
+    RefitBVH(triangles) {
+        this.triangles = triangles;
+        for (let i = this.nodesUsed - 1; i >= 0; i--) {
+            if (i !== 1) {
+                const node = this.bvhNode[i];
+                if (node.isLeaf()) {
+                    // adjust bounds to contained triangles for leaf nodes
+                    this.UpdateNodeBounds(i);
+                    continue;
+                }
+                // adjust boudns to child node bounds in interior nodes
+                const leftChild = this.bvhNode[node.leftFirst];
+                const rightChild = this.bvhNode[node.leftFirst + 1];
+                vec3MinInPlace(leftChild.aabbMin, rightChild.aabbMin, node.aabbMin);
+                vec3MaxInPlace(leftChild.aabbMax, rightChild.aabbMax, node.aabbMax);
+            }
+        }
+    }
+
+    /**
+     * Calculates the shortest distance between the origin of a ray and its intersection with the mesh
+     *
+     * @param {Ray} [ray] - The ray to be traced
+     * @param {number} [nodeIdx] - The index of the node to begin the raycast at (default is zero: the root)
+     * @ignore
+     */
+    IntersectBVH(ray, nodeIdx = 0) {
+        let node = this.bvhNode[nodeIdx];
+        const stack = [];
+        let stackPtr = 0;
+        while (true) {
+            if (node.isLeaf()) {
+                for (let i = 0; i < node.triCount; i++) {
+                    const dist = this.triangles[this.triIdx[node.leftFirst + i]].intersectWithRay(ray);
```

Comment:
You can do `const {triangles, triIdx, bvhNode} = this;` at top and prevent all these property lookups again and again and make the code shorter and faster.